strawberryfields/gbs/data.py

```diff
@@ -0,0 +1,211 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+GBS Datasets
+============
+
+**Module name:** :mod:`strawberryfields.gbs.data`
+
+.. currentmodule:: strawberryfields.gbs.data
+
+This module provides access to pre-calculated datasets of simulated GBS samples generated from a
+range of input graphs. Each dataset corresponds to a set of samples generated from a fixed graph
+with corresponding values for:
+
+- ``n_mean``: mean number of photons in the GBS device
+
+- ``n_max``: maximum number of photons allowed in any sample
+
+-  ``threshold``: flag to indicate whether samples are generated with threshold detection or
+   with photon-number-resolving detectors.
+
+- ``n_samples``: total number of samples in the dataset
+
+- ``modes``: number of modes in the GBS device or, equivalently, number of nodes in the graph
+
+Datasets are available as classes. The following graphs and datasets are provided:
+
+- **Planted graph**: A 30-node graph with a dense 10-node subgraph planted inside
+  :cite:`arrazola2018using`.
+
+  .. image:: ../../_static/planted.png
+        :align: center
+        :width: 300px
+
+  Provided by the :class:`Planted` Dataset with attributes ``n_mean = 6``, ``n_max = 14``,
+  ``threshold = True``, ``n_samples = 50000``, ``modes = 30``.
+
+Loading data
+^^^^^^^^^^^^
+
+We use the :class:`Planted` class as an example to show how to interact with the datasets. Datasets
+can be loaded by running:
+
+>>> data = Planted()
+
+Simply use indexing and slicing to access samples from the dataset:
+
+>>> sample_3 = data[3]
+>>> samples = data[:10]
+
+Datasets also contain metadata relevant to the GBS setup:
+
+>>> data.n_mean
+6
+
+>>> len(data)
+50000
+
+The number of photons or clicks in each sample is available using the :meth:`Dataset.counts` method:
+
+>>> data.counts()
+[2, 0, 8, 11, ... , 6]
+
+For example, we see that the ``data[3]`` sample had 11 clicks.
+
+Datasets
+^^^^^^^^
+
+The :class:`Dataset` class provides the base functionality from which all datasets such as
+:class:`Planted` inherit.
+
+.. autosummary::
+    Dataset
+    Planted
+
+Code details
+^^^^^^^^^^^^
+"""
+# pylint: disable=unnecessary-pass
+from abc import ABCMeta, abstractmethod
+
+import pkg_resources
+import numpy as np
+import scipy
+
+DATA_PATH = pkg_resources.resource_filename("strawberryfields", "gbs/data") + "/"
+
+
+class Dataset(metaclass=ABCMeta):

```

Comment:
Note that while this was the recommended way of creating an abstract class in Python<3.4, since 3.4, the better approach has been to do
```python
from abc import ABC

class Dataset(ABC):
```

i.e., inheritance over defining the metaclass.

Reply:
Nice thanks Josh! Done.